src/transformers/models/blip/modeling_blip.py

```diff
@@ -1191,10 +1204,17 @@ def forward(
 
         question_embeds = question_embeds[0] if not return_dict else question_embeds.last_hidden_state
 
-        if decoder_input_ids is None:
-            decoder_input_ids = torch.LongTensor([self.decoder_bos_token_id]).repeat((batch_size, 1))
+        if labels is not None and decoder_input_ids is None:
+            # get decoder inputs from shifting lm labels to the right - this is used in training mode
+            decoder_input_ids = self._shift_right(labels)
+            # replace possible -100 values in labels by `pad_token_id`
+            labels = labels.masked_fill(labels == self.decoder_pad_token_id, -100)
+        elif decoder_input_ids is None:
+            # by default use BOS token as decoder_input_ids
+            decoder_input_ids = torch.LongTensor([self.decoder_start_token_id]).repeat((batch_size, 1))
```

Comment:
I'm not sure there's a need for this.

This is handled by the generate method automatically, which will set the `decoder_input_ids` appropriately.

See also BART and T5 who don't have these lines.

Reply:
This can be removed indeed
For consistency with the original implementation, I propose to add a safety checker to check that either a label or `decoder_input_ids` are always passed: https://github.com/salesforce/BLIP/blob/3a29b7410476bf5f2ba0955827390eb6ea1f4f9d/models/blip_vqa.py#L46 
When calling the forward pass it seems that `labels` (i.e. `answer` on the source code) is always expected.